fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,585 @@
+            for (PipId pip : ctx->getPipsDownhill(node_to_expand.wire)) {
+                if (ctx->is_pip_synthetic(pip))
+                    continue;
+
+                handle_expansion_node(ctx, prev_wire, pip, node_to_expand, nodes_to_expand, bels, direction);
+            }
+        } else {
+            NPNR_ASSERT(direction == CLUSTER_UPHILL_DIR);
+            for (PipId pip : ctx->getPipsUphill(node_to_expand.wire)) {
+                if (ctx->is_pip_synthetic(pip))
+                    continue;
+
+                handle_expansion_node(ctx, prev_wire, pip, node_to_expand, nodes_to_expand, bels, direction);
+            }
+        }
+    }
+
+    return bels;
+}
+
+CellInfo *Arch::getClusterRootCell(ClusterId cluster) const
+{
+    NPNR_ASSERT(cluster != ClusterId());
+    return clusters.at(cluster).root;
+}
+
+bool Arch::getClusterPlacement(ClusterId cluster, BelId root_bel,
+                               std::vector<std::pair<CellInfo *, BelId>> &placement) const
+{
+    const Context *ctx = getCtx();
+    const Cluster &packed_cluster = clusters.at(cluster);
+
+    CellInfo *root_cell = getClusterRootCell(cluster);
+    if (!ctx->isValidBelForCellType(root_cell->type, root_bel))
+        return false;
+
+    BelId next_bel;
+
+    // Place cluster
+    for (CellInfo *cluster_node : packed_cluster.cluster_nodes) {
+        if (cluster_node == root_cell) {
+            next_bel = root_bel;
+        } else {
+            // Find next chained cluster node
+            auto &cluster_data = cluster_info(chip_info, packed_cluster.index);
+
+            IdString next_bel_pin(cluster_data.chainable_ports[0].bel_source);
+            WireId next_bel_pin_wire = ctx->getBelPinWire(next_bel, next_bel_pin);
+            next_bel = BelId();
+            for (BelId bel :
+                 find_cluster_bels(ctx, next_bel_pin_wire, CLUSTER_DOWNHILL_DIR, /*out_of_site_expansion=*/true)) {
+                if (ctx->isValidBelForCellType(cluster_node->type, bel)) {
+                    next_bel = bel;
+                    break;
+                }
+            }
+
+            if (next_bel == BelId())
+                return false;
+        }
+
+        // Build a cell to bell mapping required to find BELs connected to the cluster ports.
+        dict<IdString, std::vector<IdString>> cell_bel_pins;
+
+        int32_t mapping = bel_info(chip_info, next_bel).pin_map[get_cell_type_index(cluster_node->type)];
+        NPNR_ASSERT(mapping >= 0);
+
+        const CellBelMapPOD &cell_pin_map = chip_info->cell_map->cell_bel_map[mapping];
+        for (const auto &pin_map : cell_pin_map.common_pins) {
+            IdString cell_pin(pin_map.cell_pin);
+            IdString bel_pin(pin_map.bel_pin);
+
+            cell_bel_pins[cell_pin].push_back(bel_pin);
+        }
+
+        placement.emplace_back(cluster_node, next_bel);
+
+        for (auto port : cluster_node->ports) {
+            log_info("      - %s\n", port.first.c_str(ctx));
```

Comment:
this should be removed, or gated behind one of the debug flags and/or a macro

Reply:
This was actually some debug prints as CI is currently failing and I wanted to get some more data on why. Removed this log